feat: add versioning to transaction(input,output,kernel)

Cifko opened the pull request:
Description
---
Add versioning to Transactions.

How Has This Been Tested?
---
Some cucumber tests.

Cifko committed:
feat: add versioning to transaction(input,output,kernel)

hansieodendaal commented on base_layer/core/src/transactions/transaction/output_features.rs:
```diff
@@ -43,9 +44,26 @@ use crate::{
     },
 };
 
+#[derive(Debug, Clone, Hash, PartialEq, Deserialize, Serialize, Eq)]
+#[repr(u8)]
+pub enum OutputFeaturesVersion {
+    DIBBLER = 0,
+}
+
+impl TryFrom<u8> for OutputFeaturesVersion {
+    type Error = String;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(OutputFeaturesVersion::DIBBLER),
+            _ => Err("Unknown version!".to_string()),
+        }
+    }
+}
```
I think this should go in a separate file

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_input.rs:
```diff
@@ -56,11 +58,29 @@ use crate::{
     },
 };
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionInputVersion {
+    DIBBLER = 0,
+}
+
+impl TryFrom<u8> for TransactionInputVersion {
+    type Error = String;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(TransactionInputVersion::DIBBLER),
+            _ => Err("Unknown version!".to_string()),
+        }
+    }
+}
+
```
I think this should go in a separate file

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_kernel.rs:
```diff
@@ -39,13 +40,30 @@ use crate::transactions::{
     transaction_protocol::{build_challenge, TransactionMetadata},
 };
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionKernelVersion {
+    DIBBLER = 0,
+}
+
+impl TryFrom<u8> for TransactionKernelVersion {
+    type Error = String;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(TransactionKernelVersion::DIBBLER),
+            _ => Err("Unknown version!".to_string()),
+        }
+    }
+}
```
I think this should go in a separate file

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_output.rs:
```diff
@@ -70,11 +71,29 @@ use crate::{
     },
 };
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionOutputVersion {
+    DIBBLER = 0,
+}
+
+impl TryFrom<u8> for TransactionOutputVersion {
+    type Error = String;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(TransactionOutputVersion::DIBBLER),
+            _ => Err("Unknown version!".to_string()),
+        }
+    }
+}
+
```
I think this should go in a separate file

hansieodendaal commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs:
```diff
@@ -756,8 +756,10 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
         let start = Instant::now();
         let conn = self.database_connection.get_pooled_connection()?;
         let acquire_lock = start.elapsed();
+        println!("{}:{}", file!(), line!());
```
Was print this perhaps for debugging?

hansieodendaal commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs:
```diff
@@ -756,8 +756,10 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
+        println!("{}:{}", file!(), line!());
 
         let outputs_to_be_received = OutputSql::index_status(OutputStatus::ShortTermEncumberedToBeReceived, &conn)?;
+        println!("{}:{}", file!(), line!());
```
Print?

hansieodendaal commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs:
```diff
@@ -768,7 +770,9 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
             )?;
         }
 
+        println!("{}:{}", file!(), line!());
```
Print?

Cifko replied:
Forgot some debug ;-)

hansieodendaal commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs:
```diff
@@ -768,7 +770,9 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
 
+        println!("{}:{}", file!(), line!());
         let outputs_to_be_spent = OutputSql::index_status(OutputStatus::ShortTermEncumberedToBeSpent, &conn)?;
+        println!("{}:{}", file!(), line!());
```
Print?

hansieodendaal commented on base_layer/wallet/src/output_manager_service/storage/sqlite_db/mod.rs:
```diff
@@ -787,6 +791,7 @@ impl OutputManagerBackend for OutputManagerSqliteDatabase {
                 start.elapsed().as_millis()
             );
         }
+        println!("{}:{}", file!(), line!());
```
Print?

hansieodendaal commented on base_layer/core/src/transactions/transaction/output_features.rs:
```diff
@@ -43,9 +44,26 @@ use crate::{
+pub enum OutputFeaturesVersion {
+    DIBBLER = 0,
+}
+
```
I think we need an `OutputFeaturesVersion::GetCurrentVersion()` implementation here.

hansieodendaal commented on base_layer/core/src/transactions/transaction/output_features.rs:
```diff
@@ -63,6 +81,53 @@ impl OutputFeatures {
+        sidechain_checkpoint: Option<SideChainCheckpointFeatures>,
+    ) -> OutputFeatures {
+        OutputFeatures::new(
+            OutputFeaturesVersion::DIBBLER,
```
```suggestion
            OutputFeaturesVersion::GetCurrentVersion(),
```

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_input.rs:
```diff
@@ -56,11 +58,29 @@ use crate::{
+pub enum TransactionInputVersion {
+    DIBBLER = 0,
+}
+
```
I think we need a `TransactionInputVersion::GetCurrentVersion()` implementation here.

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_input.rs:
```diff
@@ -71,19 +91,42 @@ pub struct TransactionInput {
+        script_signature: ComSignature,
+    ) -> TransactionInput {
+        TransactionInput::new(
+            TransactionInputVersion::DIBBLER,
```
```suggestion
            TransactionInputVersion::GetCurrentVersion(),
```

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_input.rs:
```diff
@@ -94,17 +137,18 @@ impl TransactionInput {
                 script,
                 sender_offset_public_key,
                 covenant,
+                version: TransactionOutputVersion::DIBBLER,
```
```suggestion
                version: TransactionOutputVersion::GetCurrentVersion(),
```

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_kernel.rs:
```diff
@@ -62,6 +80,41 @@ pub struct TransactionKernel {
+        excess_sig: Signature,
+    ) -> TransactionKernel {
+        TransactionKernel::new(
+            TransactionKernelVersion::DIBBLER,
```
```suggestion
            TransactionKernelVersion::GetCurrentVersion(),
```

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_output.rs:
```diff
@@ -115,6 +136,27 @@ impl TransactionOutput {
+        covenant: Covenant,
+    ) -> TransactionOutput {
+        TransactionOutput::new(
+            TransactionOutputVersion::DIBBLER,
```
```suggestion
            TransactionOutputVersion::GetCurrentVersion(),
```

hansieodendaal commented on base_layer/core/src/transactions/transaction/unblinded_output.rs:
```diff
@@ -99,6 +103,33 @@ impl UnblindedOutput {
+        covenant: Covenant,
+    ) -> Self {
+        Self::new(
+            TransactionOutputVersion::DIBBLER,
```
```suggestion
            TransactionOutputVersion::GetCurrentVersion(),
```

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_kernel.rs:
```diff
@@ -39,13 +40,30 @@ use crate::transactions::{
+pub enum TransactionKernelVersion {
+    DIBBLER = 0,
+}
+
```
I think we need a `TransactionKernelVersion::GetCurrentVersion()` implementation here.

hansieodendaal commented on base_layer/core/src/transactions/transaction/transaction_output.rs:
```diff
@@ -70,11 +71,29 @@ use crate::{
+pub enum TransactionOutputVersion {
+    DIBBLER = 0,
+}
+
```
I think we need a `TransactionOutputVersion::GetCurrentVersion()` implementation here.

hansieodendaal reviewed (commented):
Hi there,

Just some comments to consider.

Thanks.

sdbondi commented on base_layer/core/src/transactions/transaction/output_features.rs:
```diff
@@ -43,9 +44,26 @@ use crate::{
+#[derive(Debug, Clone, Hash, PartialEq, Deserialize, Serialize, Eq)]
+#[repr(u8)]
+pub enum OutputFeaturesVersion {
+    DIBBLER = 0,
```
Versions don't correspond with networks. Suggest `V1`, `V2` etc

sdbondi commented on applications/tari_app_grpc/src/conversions/transaction_output.rs:
```diff
@@ -56,15 +56,16 @@ impl TryFrom<grpc::TransactionOutput> for TransactionOutput {
         let covenant = Covenant::consensus_decode(&mut output.covenant.as_slice()).map_err(|err| err.to_string())?;
-        Ok(Self {
+        Ok(Self::new(
+            TransactionOutputVersion::try_from(output.version as u8)?,
```
Should check for u8 overflows
```suggestion
            TransactionOutputVersion::try_from(u8::try_from(output.version).map_err(|_| "Invalid version: overflowed u8"))?,
```

Commit:
Merge branch 'development' into add-version-simple

stringhandler commented on base_layer/core/src/transactions/transaction/output_features_version.rs:
```diff
@@ -0,0 +1,26 @@
+#[derive(Debug, Clone, Hash, PartialEq, Deserialize, Serialize, Eq)]
+#[repr(u8)]
+pub enum OutputFeaturesVersion {
+    V1 = 0,
```
```suggestion
    V0 = 0,
```

stringhandler commented on base_layer/core/src/transactions/transaction/transaction_input_version.rs:
```diff
@@ -0,0 +1,26 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionInputVersion {
+    V1 = 0,
```
```suggestion
    V0 = 0,
```

stringhandler commented on base_layer/core/src/transactions/transaction/transaction_kernel_version.rs:
```diff
@@ -0,0 +1,25 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum TransactionKernelVersion {
+    V1 = 0,
```
```suggestion
    V0 = 0,
```

SWvheerden commented on base_layer/core/src/blocks/genesis_block.rs:
```diff
@@ -219,29 +219,29 @@ fn get_dibbler_genesis_block_raw() -> Block {
         vec![],
-        vec![TransactionOutput {
-            features: OutputFeatures {
+        vec![TransactionOutput::new_current_version(
```
You have changed the hash of the output, you need to update the MMR in the header of the genesis block as well.

SWvheerden commented on base_layer/core/src/transactions/transaction/mod.rs:
```diff
@@ -92,13 +100,15 @@ pub(super) fn hash_output(
         // .chain(range proof) // See docs as to why we exclude this
         .chain(script.as_bytes())
         .chain(covenant.to_consensus_bytes())
+        .chain((version as u8).to_le_bytes())
```
Look at the comment in the genesis block, you need to update the MMR in the header.

SWvheerden reviewed (changes requested):
This looks good, but you need to remember to update the genesis block header as well as the MMR has now changed.

Cifko committed:
recalculate genesis block

Cifko commented:
I've also forgot to add version to the TransactionInput hash. Added that as well. The genesis block is updated.

Cifko committed:
fix clippy

Cifko commented:
Added `Clippy.toml` for the number of arguments (set to 15) and removed the `#[allow(clippy::too_many_arguments)]`. This is only done for the `base_layer/core` crate.

Cifko committed:
clippy
